src/fastfeedparser/main.py

```diff
@@ -44,640 +47,571 @@ def parse(source):
         request = Request(
             source,
-            method='GET',
+            method="GET",
             headers={
-                'Accept-Encoding': 'gzip, deflate',
-                'User-Agent': 'fastfeedparser (+https://github.com/kagisearch/fastfeedparser)',
-            }
+                "Accept-Encoding": "gzip, deflate",
+                "User-Agent": "fastfeedparser (+https://github.com/kagisearch/fastfeedparser)",
+            },
         )
         opener = build_opener(HTTPRedirectHandler(), HTTPErrorProcessor())
         with opener.open(request, timeout=30) as response:
             response.begin()
-            content = response.read()
-            content_encoding = response.headers.get('Content-Encoding')
-            if content_encoding == 'gzip':
+            content: bytes = response.read()
+            content_encoding = response.headers.get("Content-Encoding")
+            if content_encoding == "gzip":
                 content = gzip.decompress(content)
-            elif content_encoding == 'deflate':
+            elif content_encoding == "deflate":
                 content = zlib.decompress(content, -zlib.MAX_WBITS)
-            xml_content = content
+            content_charset = response.headers.get_content_charset()
+            xml_content = (
+                content.decode(content_charset) if content_charset else content
+            )
     else:
         xml_content = source
 
+    # Ensure we have bytes for lxml
+    if isinstance(xml_content, str):
+        xml_content = xml_content.encode("utf-8", errors="replace")
+
     # Handle empty content
-    if isinstance(xml_content, str) and not xml_content.strip():
+    if not xml_content.strip():
         raise ValueError("Empty content")
 
-    # Handle decoding if content is bytes
-    if isinstance(xml_content, bytes):
-        encodings = ["utf-8", "iso-8859-1", "windows-1252"]
-        decoded = None
-        for encoding in encodings:
-            try:
-                decoded = xml_content.decode(encoding)
-                break
-            except UnicodeDecodeError:
-                continue
-        if decoded is None:
-            raise ValueError("Could not decode content with any supported encoding")
-        xml_content = decoded
-
-    # Ensure we have bytes for lxml
-    xml_content = xml_content.encode("utf-8", errors="replace")
-
-    parser = etree.XMLParser(recover=True)
+    parser = etree.XMLParser(
+        ns_clean=True,
+        recover=True,
+        collect_ids=False,
+        resolve_entities=False,
+    )
     try:
         root = etree.fromstring(xml_content, parser=parser)
     except etree.XMLSyntaxError as e:
         raise ValueError(f"Failed to parse XML content: {str(e)}")
-
-    # Check if root is None
     if root is None:
         raise ValueError("Failed to parse XML content: root element is None")
 
-    namespaces = root.nsmap
-
-    # Clean up the XML tree
-    for element in root.iter():
```

Comment:
BREAKING: I don't know the historical reason for removing null characters, but afaik XML cannot contain null chars (because it would make it noncompliant) and iterating over all elements is a notable performance hit.